flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/TestingExecutor.java

```diff
@@ -0,0 +1,204 @@
+import java.util.Map;
+
+/**
+ * A customizable {@link Executor} for testing purposes.
+ */
+class TestingExecutor implements Executor {
+
+	int numCancelCalls = 0;
+
+	int numRetrieveResultChancesCalls = 0;
+	private final List<SupplierWithException<TypedResult<List<Tuple2<Boolean, Row>>>, SqlExecutionException>> resultChanges;
+
+	int numSnapshotResultCalls = 0;
+	private final List<SupplierWithException<TypedResult<Integer>, SqlExecutionException>> snapshotResults;
+
+	int numRetrieveResultPageCalls = 0;
+	private final List<SupplierWithException<List<Row>, SqlExecutionException>> resultPages;
+
+	int numUseCatalogCalls = 0;
+	private final BiConsumerWithException<String, String, SqlExecutionException> useCatalogConsumer;
+
+	int numUseDatabaseCalls = 0;
+	private BiConsumerWithException<String, String, SqlExecutionException> useDatabaseConsumer;
+
+	TestingExecutor(
+			List<SupplierWithException<TypedResult<List<Tuple2<Boolean, Row>>>, SqlExecutionException>> resultChanges,
+			List<SupplierWithException<TypedResult<Integer>, SqlExecutionException>> snapshotResults,
+			List<SupplierWithException<List<Row>, SqlExecutionException>> resultPages,
+			BiConsumerWithException<String, String, SqlExecutionException> useCatalogConsumer,
+			BiConsumerWithException<String, String, SqlExecutionException> useDatabaseConsumer) {
+		this.resultChanges = resultChanges;
+		this.snapshotResults = snapshotResults;
+		this.resultPages = resultPages;
+		this.useCatalogConsumer = useCatalogConsumer;
+		this.useDatabaseConsumer = useDatabaseConsumer;
+	}
+
+	@Override
+	public void cancelQuery(String sessionId, String resultId) throws SqlExecutionException {
+		numCancelCalls++;
+	}
+
+	@Override
+	public TypedResult<List<Tuple2<Boolean, Row>>> retrieveResultChanges(String sessionId, String resultId) throws SqlExecutionException {
+		return resultChanges.get(Math.min(numRetrieveResultChancesCalls++, resultChanges.size() - 1)).get();
+	}
+
+	@Override
+	public List<Row> retrieveResultPage(String resultId, int page) throws SqlExecutionException {
+		return resultPages.get(Math.min(numRetrieveResultPageCalls++, resultPages.size() - 1)).get();
+	}
+
+	@Override
+	public TypedResult<Integer> snapshotResult(String sessionId, String resultId, int pageSize) throws SqlExecutionException {
+		return snapshotResults.get(Math.min(numSnapshotResultCalls++, snapshotResults.size() - 1)).get();
+	}
+
+	@Override
+	public void useCatalog(String sessionId, String catalogName) throws SqlExecutionException {
+		numUseCatalogCalls++;
+		useCatalogConsumer.accept(sessionId, catalogName);
+	}
+
+	@Override
+	public void useDatabase(String sessionId, String databaseName) throws SqlExecutionException {
+		numUseDatabaseCalls++;
+		useDatabaseConsumer.accept(sessionId, databaseName);
+	}
+
+	@Override
+	public void start() throws SqlExecutionException {
+	}
+
+	@Override
+	public String openSession(SessionContext session) throws SqlExecutionException {
+		return null;
```

Comment:
Is this a valid return value? Maybe it is better to throw an `UnsupportedOperationException` instead of returning `null`.